zebra-chain/src/block/header.rs

```diff
@@ -81,11 +81,36 @@ impl BlockHeader {
 
         Ok(())
     }
+
+    /// Check if `self.time` is less than or equal to
+    /// 2 hours in the future, according to the node's local clock (`now`).
+    ///
+    /// This is a non-deterministic rule, as clocks vary over time, and
+    /// between different nodes.
+    ///
+    /// "In addition, a full validator MUST NOT accept blocks with nTime
+    /// more than two hours in the future according to its clock. This
+    /// is not strictly a consensus rule because it is nondeterministic,
+    /// and clock time varies between nodes. Also note that a block that
+    /// is rejected by this rule at a given point in time may later be
+    /// accepted."[S 7.5][7.5]
+    ///
+    /// [7.5]: https://zips.z.cash/protocol/protocol.pdf#blockheader
+    pub fn is_time_valid_local_clock(&self, now: DateTime<Utc>) -> Result<(), Error> {
+        let two_hours_in_the_future = now
+            .checked_add_signed(Duration::hours(2))
+            .ok_or("overflow when calculating 2 hours in the future")?;
+        if self.time <= two_hours_in_the_future {
+            Ok(())
+        } else {
+            Err("block header time is more than 2 hours in the future")?
+        }
+    }
 }
 
 #[non_exhaustive]
 #[derive(Debug, thiserror::Error)]
-pub enum Error {
+pub enum EquihashError {
```

Comment:
non-blocking on this PR, but just something I noticed: I think we might be able to get rid of this error type entirely, since serialization is infallible mod the underlying writer, and presumably the verification code serializes into a `Vec` (which shouldn't fail except by OOM).

Reply:
I think we would just need `is_equihash_solution_valid` to return `Result<(), equihash::Error>` and use an `.expect` on the `zcash_serialize()`.

Reply:
I opened #658 to follow this up.